language-plutus-core/src/Language/PlutusCore/Instance/Eq.hs

```diff
@@ -0,0 +1,34 @@
+-- | 'Eq' instances for core data types.
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Language.PlutusCore.Instance.Eq
+    ( Global (..)
+    ) where
+
+import           PlutusPrelude
+
+import           Language.PlutusCore.Instance.Eq.Internal
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Type
+
+-- | A wrapper around any @name@ needed for the sake of its 'Eq' instance, which requires the @name@
+-- to have a 'Unique' and the 'Unique's of two names are used for performing the equality check.
+newtype Global name = Global
```

Comment:
For checking equality of abstract names elsewhere we could write `name1 ^. unique == name2 ^. unique`, but I prefer to insist on writing `Global name1 == Global name2`, so that it's clearly that the global uniqueness condition must be satisfied when this check is performed (otherwise two names can be considered equal even though they are different semantically and even their textual representations may differ and it'd really weird to return `True` in that case).

Name handling is quite tricky, so I like to be explicit about it.

Reply:
Ah I see, we've got rid of the `Eq` instance for `Name`.

I guess I'd probably just have written a function `globalEq :: Name -> Name -> Bool` or whatever.

Reply:
Also the only uses are for `VarDecl`s and the like, where it's not that obvious that they're relying on global uniqueness!

Reply:
Hm. Well yeah, I guess it's better to get rid of `Eq` instance from `MkPlc` then and of this `Global` thing as well. I'll do that.

Reply:
> Ah I see, we've got rid of the `Eq` instance for `Name`.

We actually didn't. It's still there and relies on uniques. We need it for `Eq` instances for errors. But those instances are not super important. Anyway, I don't want to refactor all of that in this PR.

Reply:
> Well yeah, I guess it's better to get rid of `Eq` instance from `MkPlc` then and of this `Global` thing as well. 

Done.